### ISSUE TITLE
Vue asset images not loading

### ISSUE THREAD

**Issue by fylzero** [laravel-mix/laravel-mix#2756]
-   Laravel Mix Version: 6.0.9
-   Node Version: 15.5.1
-   NPM Version: 7.3.0
-   OS: Mac OS Big Sur 11.1

### Description:
After upgrading to Laravel Mix 6 my build runs successfully but images referenced from the assets folder no longer are loading.

### Steps To Reproduce:
Running an existing Laravel v8.21.0 project with Laravel/UI Vue auth scaffolding, after upgrading Laravel Mix to version 6 images referenced from the assets folder in Vue don't err out on build but display as broken images in the browser.

Referencing images as such:

```
<img src="../assets/logo.png" alt="Logo">
```

When I inspect the image I see this:

```
<img src="[object Module]" alt="Logo">
```

Any help would be greatly appreciated.  Thanks!

**Comment by Aiwings** [comment]
Hi, 
I faced a similar issue yesterday. 
The assets were copied successfully but the src path were not correctly rendered (object instead of a string)

It turned out this was caused by the embedded file-loader, which by default output files with ES6 syntax.
So I had to override the original configuration
`mix.override(webpackConfig => {
   // Firstly try to find the rule(s) where the file-loader is implyed
    console.log(webpackConfig.modules.rules)
   // Then modify the loader configuration
    webpackConfig.module.rules[2].use[0].options.esModule = false;
});`
Hope it can help.

**Comment by fylzero** [comment]
@Aiwings Thanks for the response.

I was also able to get the image loaded by doing this:

```
<img :src="require('./assets/logo.png').default" alt="Logo">
```

Both of these solutions feel kinda ugly to me tbh.

**Comment by tinyfly** [comment]
Apparently its a bug with vue-loader which Mix uses. Here is my workaround solution. This only sets esModule to false on the images file-loader.

```javascript
mix.override(webpackConfig => {
  // BUG: vue-loader doesn't handle file-loader's default esModule:true setting properly causing
  // <img src="[object module]" /> to be output from vue templates.
  // WORKAROUND: Override mixs and turn off esModule support on images.
  // FIX: When vue-loader fixes their bug AND laravel-mix updates to the fixed version
  // this can be removed
  webpackConfig.module.rules.forEach(rule => {
    if (rule.test.toString() === '/(\\.(png|jpe?g|gif|webp)$|^((?!font).)*\\.svg$)/') {
      if (Array.isArray(rule.use)) {
        rule.use.forEach(ruleUse => {
          if (ruleUse.loader === 'file-loader') {
            ruleUse.options.esModule = false;
          }
        });
      }
    }
  });
});
```

**Comment by ethanclevenger91** [comment]
@tinyfly you have a link to an open issue for that?

**Comment by tinyfly** [comment]
@ethanclevenger91 Here you go: https://github.com/vuejs/vue-loader/issues/1612

**Comment by Bouhnosaure** [comment]
> Apparently its a bug with vue-loader which Mix uses. Here is my workaround solution. This only sets esModule to false on the images file-loader.
> 
> ```js
> mix.override(webpackConfig => {
>   // BUG: vue-loader doesn't handle file-loader's default esModule:true setting properly causing
>   // <img src="[object module]" /> to be output from vue templates.
>   // WORKAROUND: Override mixs and turn off esModule support on images.
>   // FIX: When vue-loader fixes their bug AND laravel-mix updates to the fixed version
>   // this can be removed
>   webpackConfig.module.rules.forEach(rule => {
>     if (rule.test.toString() === '/(\\.(png|jpe?g|gif|webp)$|^((?!font).)*\\.svg$)/') {
>       if (Array.isArray(rule.use)) {
>         rule.use.forEach(ruleUse => {
>           if (ruleUse.loader === 'file-loader') {
>             ruleUse.options.esModule = false;
>           }
>         });
>       }
>     }
>   });
> });
> ```

I had this issue when i've updated laravel mix from 5 to 6 

Got some [object Module] urls for thirds libraries like trumbowyg or remixicons and this solved the issue

Thanks, i hope there will be a fix for this or an option to set the esModule option for file-loader !

**Comment by fylzero** [comment]
@tinyfly I finally got around to implementing your solution and it works great!  Thank you for that and for linking to the issue in vue-loader.

**Comment by FlavioSB** [comment]
`<img :src="require(`@assets/img/test/test.png`).default" alt="sideimg">`
Was the only thing which worked for me. Without the @assets (custom alias in my project) it would not get copied to the public folder and without the `.default` it would show `[object Module]` in the html and not render at all.

But I dont feel like this is a definitive solution, it shouldnt be necessary to add `.default` OR use `require` at all in my opinion.

**Comment by 321zeno** [comment]
Overriding the file-loader via `mix.webpackConfig()` worked for me:

```js
mix.webpackConfig(webpack => {
    return {
        module: {
            rules: [{
                test: /(\.(png|jpe?g|gif|webp)$|^((?!font).)*\.svg$)/,
                use: [{
                    loader: require.resolve('file-loader'),
                    options: {
                        esModule: false
                    }
                }]
            }]
        }
    };
});
```

The `test` regex has to match the default Webpack config in order to override the rule. I used `mix.dump()` to find out the 
default Webpack config.


The config set using `mix.override()` seemed to be ignored in Mix 6.

**Comment by tinyfly** [comment]
> Overriding the file-loader via `mix.webpackConfig()` worked for me:

@321zeno I like this solution better than mine. I'm going to give it a shot. Thanks!

**Comment by fylzero** [comment]
@321zeno Implemented your solution, works great!

**Comment by fylzero** [comment]
@321zeno This solution randomly stopped working for me today.  I initially thought this was due to Laravel Mix updating from 6.0.31 to 6.0.39 but reverting didn't seem to help.  I also thought maybe file-loader was removed as a sub-dependency somewhere but that doesn't seem to be the case.  Anyone else having this issue?

**Comment by loilo** [comment]
Yes, the same happened for me (although I never had it working before).
The reason seems to be that Mix [added support for AVIF images](https://github.com/laravel-mix/laravel-mix/blame/a2e8c5f14156b1495efe4bb67a4256fb24427460/src/builder/webpack-rules.js#L23) in the loader, altering the regular expression used. Updating the override accordingly made the fix work again:

```diff
mix.webpackConfig(webpack => {
    return {
        module: {
            rules: [{
-                test: /(\.(png|jpe?g|gif|webp)$|^((?!font).)*\.svg$)/,
+                test: /(\.(png|jpe?g|gif|webp|avif)$|^((?!font).)*\.svg$)/,
                use: [{
                    loader: require.resolve('file-loader'),
                    options: {
                        esModule: false
                    }
                }]
            }]
        }
    };
});
```

**Comment by Aslam97** [comment]
@loilo and mix also updated the loader from `loader: 'file-loader'` using path `project/node_modules/file-loader/dist/cjs.js'`
for me instead of comparing regex strings or writing your own rules i used this solution

```js
const mix = require('laravel-mix')

mix.extend('disabledEsModule', (config) => {
  config.module.rules.forEach((rule) => {
    let regex = new RegExp(rule.test)

    // why .jpg? because it's one of those file extensions that is someone never left him behind.
    if (regex.test('.jpg')) {
      rule.use.forEach((use) => {
        if (use.loader.includes('file-loader')) {
          Object.assign(use.options, { esModule: false })
        }
      })
    }
  })
})
```

or use this package for more simple https://www.npmjs.com/package/laravel-mix-fileloader

**Comment by Siada-Apius** [comment]
I have faced with the same issue during update from v4 to v6. @321zeno thank you!

**Comment by AbdullahAgsar** [comment]
> @fylzero Yes, the same happened for me (although I never had it working before). The reason seems to be that Mix [added support for AVIF images](https://github.com/laravel-mix/laravel-mix/blame/a2e8c5f14156b1495efe4bb67a4256fb24427460/src/builder/webpack-rules.js#L23) in the loader, altering the regular expression used. Updating the override accordingly made the fix work again:
> 
> ```diff
> mix.webpackConfig(webpack => {
>     return {
>         module: {
>             rules: [{
> -                test: /(\.(png|jpe?g|gif|webp)$|^((?!font).)*\.svg$)/,
> +                test: /(\.(png|jpe?g|gif|webp|avif)$|^((?!font).)*\.svg$)/,
>                 use: [{
>                     loader: require.resolve('file-loader'),
>                     options: {
>                         esModule: false
>                     }
>                 }]
>             }]
>         }
>     };
> });
> ```

it is work in my project. @Aslam97 thanks.
